File: honjo-masamune/honjo/src/interp.ts
// Honjo Masamune — Interpreter (operational semantics §6)
// Small-step evaluation over Cut-IR. Configuration = <graph state, cut-count M,
// instruction pointer>. Evaluation IS measurement: each cut event mutates state
// and strictly increments M (Thm "cut monotonicity"). This is the exact
// (reference) realisation of a cut — the back end the spec calls Target R in
// semantics, here in TypeScript for the playground; the WebGL render-pass back
// end (Target T) consumes the same IR.

import { IRProgram, IROperand, IRCond } from "./ir.js";
import {
  AtomVal, CompoundVal, PathVal, CutVal,
  individuate, bond, close, propagate,
} from "./stdlib.js";

export interface RunResult {
  cutCount: number;            // M, the program's intrinsic clock
  floor: number;
  registers: (RVal | undefined)[];
  named: Record<string, RVal>; // bound names -> values
  log: string[];               // observe / assert output
  ok: boolean;                 // false if an assert aborted
}

export type RVal = CutVal | { ty: "Scalar"; value: number; floor: number };

export class RuntimeError extends Error {
  constructor(msg: string) { super(`runtime error: ${msg}`); this.name = "RuntimeError"; }
}

export function run(ir: IRProgram): RunResult {
  const regs: (RVal | undefined)[] = [];
  const named: Record<string, RVal> = {};
  const log: string[] = [];
  let floor = Number.isNaN(ir.floor) ? 1 : ir.floor;
  let M = 0; // committed-cut count; monotone

  const regToName = new Map<number, string>();
  for (const [name, reg] of ir.nameToReg) regToName.set(reg, name);

  const getReg = (id: number): RVal => {
    const v = regs[id];
    if (v === undefined) throw new RuntimeError(`register r${id} read before write`);
    return v;
  };

  const asAtom = (id: number): AtomVal => {
    const v = getReg(id);
    if (v.ty !== "Atom") throw new RuntimeError(`expected Atom in r${id}, got ${v.ty}`);
    return v;
  };

  // resolve a numeric operand: literals, scalar regs, and measured fields.
  const numOperand = (o: IROperand): number => {
    switch (o.kind) {
      case "lit": return o.value;
      case "reg": {
        const v = getReg(o.id);
        return "value" in v ? (v as any).value : (v as any).residue ?? NaN;
      }
      case "field": return fieldValue(getReg(o.reg), o.field);
      case "name": throw new RuntimeError(`unresolved measured field '${o.name}' (no value in scope)`);
    }
  };

  // resolve a field of a cut value (delta, vacancy, valence, residue, angle...)
  const fieldValue = (v: RVal, field: string): number => {
    const anyv = v as any;
    if (field === "valence" || field === "closed") {
      if (v.ty === "Compound") return (v as CompoundVal).valenceClosed ? 1 : 0;
      if (v.ty === "Atom") return (v as AtomVal).vacancy === 0 ? 1 : 0;
    }
    if (field in anyv && typeof anyv[field] === "number") return anyv[field];
    throw new RuntimeError(`value of type ${v.ty} has no numeric field '${field}'`);
  };

  // evaluate a condition. A bare `name` operand refers to a field of the most
  // recent cut value when it matches (delta on a bond guard, etc.).
  const evalCond = (c: IRCond, ctx?: RVal): boolean => {
    const resolve = (o: IROperand): number => {
      if (o.kind === "name" && ctx) {
        // `delta`, `vacancy`, ... read off the contextual value
        try { return fieldValue(ctx, o.name); } catch { /* fall through */ }
      }
      if (o.kind === "name") {
        // closed-shell sentinel for `valence == closed`
        if (o.name === "closed") return ctx ? fieldValue(ctx, "valence") : 1;
      }
      return numOperand(o);
    };
    const L = resolve(c.left), R = resolve(c.right);
    switch (c.op) {
      case ">": return L > R;
      case "<": return L < R;
      case ">=": return L >= R;
      case "<=": return L <= R;
      case "==": return L === R;
    }
  };

  for (const ins of ir.instrs) {
    switch (ins.op) {
      case "Floor":
        if (!(ins.value > 0)) throw new RuntimeError("floor must be > 0");
        floor = ins.value;
        break;

      case "Mov":
        regs[ins.dst] = scalarFromOperand(ins.src, floor, numOperand);
        break;

      case "Ind": {
        const z = numOperand(ins.z);
        const atom = individuate(z, floor);
        regs[ins.dst] = atom;
        M += 1; // a cut event
        break;
      }

      case "Bnd": {
        const a = asAtom(ins.a), b = asAtom(ins.b);
        const bd = bond(a, b, floor);
        // guard, if present, reads `delta` off the bond just formed
        if (ins.guard && !evalCond(ins.guard, bd)) {
          // guard fails -> no bond committed (still a measured value, residue floor)
          regs[ins.dst] = { ...bd, exists: false };
          break;
        }
        regs[ins.dst] = bd;
        M += 1;
        break;
      }

      case "Cls": {
        const central = asAtom(ins.central);
        const ligs = ins.args.map((r) => asAtom(r));
        const comp = close(central, ligs, floor);
        regs[ins.dst] = comp;
        M += comp.ligands; // closure commits one cut per ligand interface
        break;
      }

      case "Prp": {
        const item = asAtom(ins.item);
        const proc = getReg(ins.process);
        if (proc.ty !== "Compound" && proc.ty !== "Path")
          throw new RuntimeError(`track process must be Compound or Path, got ${proc.ty}`);
        const admit =
          ins.admit.kind === "converge" ? "converge" as const :
          ins.admit.kind === "diverge" ? "diverge" as const :
          { holds: evalCond(ins.admit.cond, proc) };
        const path = propagate(item, proc as CompoundVal | PathVal, ins.reps ?? [], admit, floor);
        regs[ins.dst] = path;
        M += path.steps;
        break;
      }

      case "Obs": {
        const v = getReg(ins.reg);
        log.push(renderValue(v, regToName.get(ins.reg)));
        if (ins.as !== undefined) regs[ins.as] = v;
        break;
      }

      case "Assert": {
        const ctx = lastCutContext(regs);
        if (!evalCond(ins.cond, ctx)) {
          log.push(`ABORT: ${ins.message ?? "assertion failed"}`);
          return finalize(false);
        }
        break;
      }

      case "Call": {
        // module/stdlib call by name; supports the canonical verbs by qname.
        const verb = ins.name[ins.name.length - 1];
        if (verb === "individuate" || verb === "atom") {
          const z = numOperand(ins.args[0]);
          regs[ins.dst] = individuate(z, floor);
          M += 1;
        } else {
          throw new RuntimeError(`unknown call '${ins.name.join(".")}'`);
        }
        break;
      }
    }
  }

  return finalize(true);

  function finalize(ok: boolean): RunResult {
    for (const [name, reg] of ir.nameToReg) {
      const v = regs[reg];
      if (v !== undefined) named[name] = v;
    }
    return { cutCount: M, floor, registers: regs, named, log, ok };
  }
}

function scalarFromOperand(o: IROperand, floor: number, num: (o: IROperand) => number): RVal {
  if (o.kind === "lit") return { ty: "Scalar", value: o.value, floor: o.floor || floor };
  return { ty: "Scalar", value: num(o), floor };
}

function lastCutContext(regs: (RVal | undefined)[]): RVal | undefined {
  for (let i = regs.length - 1; i >= 0; i--) {
    const v = regs[i];
    if (v && v.ty !== "Scalar") return v;
  }
  return undefined;
}

export function renderValue(v: RVal, name?: string): string {
  const tag = name ? `${name} : ` : "";
  switch (v.ty) {
    case "Atom":
      return `${tag}Atom @ ${fmt(v.floor)}  Z=${v.Z} ${v.symbol}  ${v.config}  ${v.term}  vacancy=${v.vacancy}  valence=${v.valence}  residue=${fmt(v.residue)}`;
    case "Bond":
      return `${tag}Bond @ ${fmt(v.floor)}  ${v.a}~${v.b}  exists=${v.exists}  delta=${fmt(v.delta)}  shared=${v.shared}  residue=${fmt(v.residue)}`;
    case "Compound": {
      const lig = v.formula[1] > 1 ? v.ligand + v.formula[1] : v.formula[1] === 1 ? v.ligand : "";
      const formula = v.formula[0] === 2 ? v.central + "2" : v.central + lig;
      return `${tag}Compound @ ${fmt(v.floor)}  ${formula}  geometry=${v.geometry}  angle=${v.angleDeg ?? "-"}  closed=${v.valenceClosed}  residue=${fmt(v.residue)}`;
    }
    case "Path":
      return `${tag}Path @ ${fmt(v.floor)}  item=${v.item}  steps=${v.steps}  converged=${v.converged}  reps=[${v.reps.join(",")}]  amalgamation=[${v.amalgamation.join(", ")}]  residue=${fmt(v.residue)}`;
    case "Scalar":
      return `${tag}Scalar @ ${fmt(v.floor)}  ${fmt(v.value)}`;
  }
}

function fmt(n: number): string {
  if (Number.isNaN(n)) return "NaN";
  if (Number.isInteger(n)) return String(n);
  return n.toFixed(4).replace(/0+$/, "").replace(/\.$/, "");
}

File: honjo-masamune/honjo-rs/src/interp.rs
//! Honjo Masamune — interpreter (operational semantics §6).
//! Evaluation IS measurement: each cut event mutates state and strictly
//! increments the cut count M (the intrinsic clock, monotone). This is the
//! exact (reference) realisation; a cut's residue is the exact boundary weight.

use crate::ast::*;
use crate::stdlib::{self, Admit as SAdmit, AtomVal, Value};
use std::collections::HashMap;

pub struct RunResult {
    pub cut_count: i64,
    pub floor: f64,
    pub named: HashMap<String, Value>,
    pub log: Vec<String>,
    pub ok: bool,
}

struct Interp {
    floor: f64,
    m: i64, // committed-cut count; monotone
    env: HashMap<String, Value>,
    log: Vec<String>,
    // insertion order of named values, for deterministic reporting
    order: Vec<String>,
}

type IResult<T> = Result<T, String>;

impl Interp {
    fn bind(&mut self, name: &str, v: Value) {
        if !self.env.contains_key(name) {
            self.order.push(name.to_string());
        }
        self.env.insert(name.to_string(), v);
    }

    fn get_atom(&self, name: &str) -> IResult<AtomVal> {
        match self.env.get(name) {
            Some(Value::Atom(a)) => Ok(a.clone()),
            Some(other) => Err(format!("expected Atom for '{}', got {:?}", name, kind_of(other))),
            None => Err(format!("unbound name '{}'", name)),
        }
    }

    fn eval_expr(&mut self, e: &Expr) -> IResult<Value> {
        match e {
            Expr::Num { value, floor, .. } => {
                Ok(Value::Scalar { value: *value, floor: floor.unwrap_or(self.floor) })
            }
            Expr::Str { .. } => Ok(Value::Scalar { value: f64::NAN, floor: self.floor }),
            Expr::Ref { name, .. } => self
                .env
                .get(name)
                .cloned()
                .ok_or_else(|| format!("unbound name '{}'", name)),
            Expr::Cut { arg, .. } => {
                let z = self.eval_scalar(arg)?;
                let atom = stdlib::individuate(z as i64, self.floor)?;
                self.m += 1;
                Ok(Value::Atom(atom))
            }
            Expr::Bond { left, right, guard, .. } => {
                let a = self.eval_atom(left)?;
                let b = self.eval_atom(right)?;
                let bd = stdlib::bond(&a, &b, self.floor);
                if let Some(g) = guard {
                    if !self.eval_cond(g, Some(&Value::Bond(bd.clone())))? {
                        let mut dead = bd.clone();
                        dead.exists = false;
                        return Ok(Value::Bond(dead));
                    }
                }
                self.m += 1;
                Ok(Value::Bond(bd))
            }
            Expr::Close { central, args, .. } => {
                let c = self.get_atom(central)?;
                let mut ligs = Vec::new();
                for a in args {
                    ligs.push(self.eval_atom(&a.value)?);
                }
                let comp = stdlib::close(&c, &ligs, self.floor)?;
                self.m += comp.ligands;
                Ok(Value::Compound(comp))
            }
            Expr::TrackExpr { item, process, reps, admit, .. } => {
                self.eval_track(item, process, reps, admit)
            }
            Expr::Call { name, args, .. } => {
                let verb = name.last().map(|s| s.as_str()).unwrap_or("");
                if verb == "individuate" || verb == "atom" {
                    let z = self.eval_scalar(&args[0].value)?;
                    let atom = stdlib::individuate(z as i64, self.floor)?;
                    self.m += 1;
                    Ok(Value::Atom(atom))
                } else {
                    Err(format!("unknown call '{}'", name.join(".")))
                }
            }
        }
    }

    fn eval_track(
        &mut self,
        item: &str,
        process: &Expr,
        reps: &Option<Vec<String>>,
        admit: &Admit,
    ) -> IResult<Value> {
        let it = self.get_atom(item)?;
        let proc = self.eval_expr(process)?;
        let sadmit = match admit {
            Admit::Converge => SAdmit::Converge,
            Admit::Diverge => SAdmit::Diverge,
            Admit::Cond(c) => SAdmit::Cond(self.eval_cond(c, Some(&proc))?),
        };
        let reps_v = reps.clone().unwrap_or_default();
        let path = stdlib::propagate(&it, &proc, &reps_v, sadmit, self.floor)?;
        self.m += path.steps;
        Ok(Value::Path(path))
    }

    fn eval_scalar(&mut self, e: &Expr) -> IResult<f64> {
        match self.eval_expr(e)? {
            Value::Scalar { value, .. } => Ok(value),
            other => Err(format!("expected Scalar, got {:?}", kind_of(&other))),
        }
    }

    fn eval_atom(&mut self, e: &Expr) -> IResult<AtomVal> {
        match self.eval_expr(e)? {
            Value::Atom(a) => Ok(a),
            other => Err(format!("expected Atom, got {:?}", kind_of(&other))),
        }
    }

    fn eval_cond(&mut self, c: &Cond, ctx: Option<&Value>) -> IResult<bool> {
        let l = self.eval_operand(&c.left, ctx)?;
        let r = self.eval_operand(&c.right, ctx)?;
        Ok(match c.op {
            RelOp::Gt => l > r,
            RelOp::Lt => l < r,
            RelOp::Ge => l >= r,
            RelOp::Le => l <= r,
            RelOp::Eq => l == r,
        })
    }

    fn eval_operand(&mut self, e: &Expr, ctx: Option<&Value>) -> IResult<f64> {
        match e {
            // bare name: a measured field of the context value (delta, valence...)
            Expr::Ref { name, .. } => {
                if let Some(v) = self.env.get(name) {
                    return Ok(scalar_of(v));
                }
                if name == "closed" {
                    // sentinel: `valence == closed` -> 1 when closed
                    return Ok(1.0);
                }
                if let Some(cx) = ctx {
                    return field_value(cx, name);
                }
                Err(format!("unresolved measured field '{}'", name))
            }
            // dotted field W.valence
            Expr::Call { name, args, .. } if name.len() == 2 && args.is_empty() => {
                let base = self
                    .env
                    .get(&name[0])
                    .cloned()
                    .ok_or_else(|| format!("unbound '{}'", name[0]))?;
                field_value(&base, &name[1])
            }
            _ => Ok(scalar_of(&self.eval_expr(e)?)),
        }
    }

    fn exec_stmt(&mut self, s: &Stmt) -> IResult<bool> {
        match s {
            Stmt::Bind { name, value, .. } => {
                let v = self.eval_expr(value)?;
                self.bind(name, v);
            }
            Stmt::ExprStmt { expr, .. } => {
                self.eval_expr(expr)?;
            }
            Stmt::Observe { expr, as_name, .. } => {
                let v = self.eval_expr(expr)?;
                let label = expr_name(expr);
                self.log.push(render_value(&v, label.as_deref()));
                if let Some(a) = as_name {
                    self.bind(a, v);
                }
            }
            Stmt::Assert { cond, emit, .. } => {
                let ctx = self.last_cut_context();
                if !self.eval_cond(cond, ctx.as_ref())? {
                    self.log.push(format!("ABORT: {}", emit.clone().unwrap_or_else(|| "assertion failed".into())));
                    return Ok(false);
                }
            }
            Stmt::Track { item, process, reps, admit, yield_name, .. } => {
                let v = self.eval_track(item, process, reps, admit)?;
                self.bind(yield_name, v);
            }
        }
        Ok(true)
    }

    fn last_cut_context(&self) -> Option<Value> {
        for name in self.order.iter().rev() {
            if let Some(v) = self.env.get(name) {
                if !matches!(v, Value::Scalar { .. }) {
                    return Some(v.clone());
                }
            }
        }
        None
    }

    fn exec_decl(&mut self, d: &Decl) -> IResult<bool> {
        match d {
            Decl::Floor { value, .. } => {
                if *value <= 0.0 {
                    return Err("floor must be > 0".into());
                }
                self.floor = *value;
            }
            Decl::Import { .. } => {}
            Decl::Module { body, .. } => {
                for inner in body {
                    if !self.exec_decl(inner)? {
                        return Ok(false);
                    }
                }
            }
            Decl::Stmt(s) => return self.exec_stmt(s),
        }
        Ok(true)
    }
}

fn kind_of(v: &Value) -> &'static str {
    match v {
        Value::Scalar { .. } => "Scalar",
        Value::Atom(_) => "Atom",
        Value::Bond(_) => "Bond",
        Value::Compound(_) => "Compound",
        Value::Path(_) => "Path",
    }
}

fn scalar_of(v: &Value) -> f64 {
    match v {
        Value::Scalar { value, .. } => *value,
        Value::Atom(a) => a.residue,
        Value::Bond(b) => b.residue,
        Value::Compound(c) => c.residue,
        Value::Path(p) => p.residue,
    }
}

fn field_value(v: &Value, field: &str) -> Result<f64, String> {
    if field == "valence" || field == "closed" {
        match v {
            Value::Compound(c) => return Ok(if c.valence_closed { 1.0 } else { 0.0 }),
            Value::Atom(a) => return Ok(if a.vacancy == 0 { 1.0 } else { 0.0 }),
            _ => {}
        }
    }
    match (v, field) {
        (Value::Atom(a), "vacancy") => Ok(a.vacancy as f64),
        (Value::Atom(a), "residue") => Ok(a.residue),
        (Value::Bond(b), "delta") => Ok(b.delta),
        (Value::Bond(b), "shared") => Ok(b.shared as f64),
        (Value::Bond(b), "residue") => Ok(b.residue),
        (Value::Compound(c), "angle") => Ok(c.angle_deg.unwrap_or(f64::NAN)),
        (Value::Compound(c), "ligands") => Ok(c.ligands as f64),
        (Value::Compound(c), "residue") => Ok(c.residue),
        (Value::Path(p), "steps") => Ok(p.steps as f64),
        (Value::Path(p), "residue") => Ok(p.residue),
        _ => Err(format!("value {} has no numeric field '{}'", kind_of(v), field)),
    }
}

fn expr_name(e: &Expr) -> Option<String> {
    match e {
        Expr::Ref { name, .. } => Some(name.clone()),
        _ => None,
    }
}

fn fmt(n: f64) -> String {
    if n.is_nan() {
        return "NaN".into();
    }
    if n.fract() == 0.0 {
        format!("{}", n as i64)
    } else {
        let s = format!("{:.4}", n);
        let s = s.trim_end_matches('0').trim_end_matches('.');
        s.to_string()
    }
}

pub fn render_value(v: &Value, name: Option<&str>) -> String {
    let tag = name.map(|n| format!("{} : ", n)).unwrap_or_default();
    match v {
        Value::Atom(a) => format!(
            "{}Atom @ {}  Z={} {}  {}  {}  vacancy={}  valence={}  residue={}",
            tag, fmt(a.floor), a.z, a.symbol, a.config, a.term, a.vacancy, a.valence, fmt(a.residue)
        ),
        Value::Bond(b) => format!(
            "{}Bond @ {}  {}~{}  exists={}  delta={}  shared={}  residue={}",
            tag, fmt(b.floor), b.a, b.b, b.exists, fmt(b.delta), b.shared, fmt(b.residue)
        ),
        Value::Compound(c) => {
            let lig = if c.formula.1 > 1 {
                format!("{}{}", c.ligand, c.formula.1)
            } else if c.formula.1 == 1 {
                c.ligand.clone()
            } else {
                String::new()
            };
            let formula = if c.formula.0 == 2 {
                format!("{}2", c.central)
            } else {
                format!("{}{}", c.central, lig)
            };
            let angle = c.angle_deg.map(fmt).unwrap_or_else(|| "-".into());
            format!(
                "{}Compound @ {}  {}  geometry={}  angle={}  closed={}  residue={}",
                tag, fmt(c.floor), formula, c.geometry, angle, c.valence_closed, fmt(c.residue)
            )
        }
        Value::Path(p) => format!(
            "{}Path @ {}  item={}  steps={}  converged={}  reps=[{}]  amalgamation=[{}]  residue={}",
            tag, fmt(p.floor), p.item, p.steps, p.converged, p.reps.join(","), p.amalgamation.join(", "), fmt(p.residue)
        ),
        Value::Scalar { value, floor } => format!("{}Scalar @ {}  {}", tag, fmt(*floor), fmt(*value)),
    }
}

pub fn run(program: &Program) -> Result<RunResult, String> {
    let mut it = Interp {
        floor: 1.0,
        m: 0,
        env: HashMap::new(),
        log: Vec::new(),
        order: Vec::new(),
    };
    let mut ok = true;
    for d in &program.decls {
        if !it.exec_decl(d)? {
            ok = false;
            break;
        }
    }
    Ok(RunResult { cut_count: it.m, floor: it.floor, named: it.env, log: it.log, ok })
}
